docs/rfc-1-staking-contract.adoc

```diff
@@ -0,0 +1,479 @@
+case the amount of stake available after the slashing is lower than these
+authorizations.
+
+=== Limitations
+
+An absolute minimum is that existing KEEP/NU stakes should be supported in T
+network for Random Beacon, TBTC v2, and Proxy Re-Encryption. Given the fact
+supporting legacy KEEP/NU stakes forever, in every other application implemented
+for T, might not be always possible, it is expected that applications are able
+to revert authorization transactions if the delegation used in the authorization
+order does not meet the application requirements.
+
+=== Upgradeability
+
+The staking contract will be upgradeable. The exact upgradeability mechanism is
+out of the scope of this document.
+
+== Public API
+
+=== Delegating a stake
+
+==== `stake(address operator, address beneficiary, address authorizer, uint256 amount) external` 
+   
+Creates a delegation with `msg.sender` owner with the given operator,
+beneficiary, and authorizer. Transfers the given amount of T to the staking
+contract. The owner of the delegation needs to have the amount approved to
+transfer to the staking contract.
+
+==== `stakeKeep(address operator) external`
+
+Copies delegation from the legacy KEEP staking contract to T staking contract.
+No tokens are transferred. Caches the active stake amount from KEEP staking 
+contract. Can be called by anyone.
+
+==== `stakeNu(address operator, address beneficiary, address authorizer) external`
+
+Copies delegation from the legacy NU staking contract to T staking contract,
+additionally appointing beneficiary and authorizer roles. Caches the amount
+staked in NU staking contract. Can be called only by the original delegation
+owner.
+
+==== `setMinimumStakeAmount(uint256 amount) external onlyGovernance`
+
+Allows the governance to set the minimum required stake amount. This amount is
+required to protect against griefing the staking contract and individual
+applications are allowed to require higher minimum stakes if necessary.  
+
+=== Authorizing an application
+
+==== `approveApplication(address application) external onlyGovernance`
+
+Allows the governance to approve the particular application before individual
+stake authorizers are able to authorize it.
+
+==== `authorizeApplication(address operator, address application, uint256 amount) external onlyAuthorizerOf(operator)`
+
+Authorizes the particular application for the given operator. From this moment,
+the application may slash operator's stake up to the given amount. Can only be
+called by the given operator's authorizer.
+
+==== `increaseAuthorization(address operator, address application, uint256 amount) external onlyAuthorizerOf(operator)`
+
+Increases the authorization of the given operator for the given application by
+the given amount. Calls `authorizationIncreased(address operator, uint256 amount)`
+callback on the given application to notify the application. Can only be called
+by the given operator's authorizer.
```

Comment:
What do you think about to have only one method instead of `increaseAuthorization` and `authorizeApplication`? In current version of doc we have only one callback for increasing authorization. If there will be separate method for initial authorization then maybe add another callback for it?

Reply:
Good call. I think we should drop `authorizeApplication` . `increaseAuthorization` is already covered with the callback and increasing from 0 is still increasing.